pep-9999.rst

```diff
@@ -180,6 +180,17 @@ Yes::
         def do_stuff(self, y: str) -> None: ...
         doStuff = do_stuff
 
+The type of generic classes can be narrowed by annotating the ``self``
+argument of the ``__init__`` method::
+
+    class Foo(Generic[_S, _T]):
+        @overload
+        def __init__(self: Foo[str, str], type: Literal["s"]) -> None: ...
```

Comment:
My assumption is that this "self" annotation requires that the class within the annotation matches the class in which it is declared. In other words, you can't use some other (e.g. ancestor) class or protocol in the annotation. Furthermore, I think it's important to state that if the type arguments within this annotation make use of a type variable, it must be a "bare" type variable (as opposed to some complex type that includes the type variable such as `Union[List[_T], Set[_T]]`, and the type variable must be declared by the class. For example, it would be considered an error to specify type variable `_U` in the example above because `_U` is not a generic type parameter of the class `Foo`. Do those sound like reasonable limitations to place on this? I couldn't figure out how to implement it without those limitations.

Reply:
@erictraut I tried to incorporate your suggestions. I think it's fine to be strict for now, especially since we try to reflect the status quo.

Reply:
Thanks. LGTM.